### PR TITLE
Allow ".dll" extensions in BlazorWebAssemblyLazyLoad MSBuild task.

### PR DESCRIPTION
**Pull request by adrianwright109**
Addresses #24814

### CODE REVIEW

File: src/Components/WebAssembly/Sdk/src/GenerateBlazorWebAssemblyBootJson.cs [hunk 1]
```diff
@@ -149,7 +149,7 @@ public void WriteBootJson(Stream output, string entryAssemblyName)
 
         private bool IsLazyLoadedAssembly(string fileName)
         {
-            return LazyLoadedAssemblies != null && LazyLoadedAssemblies.Any(a => a.ItemSpec == fileName);
+            return LazyLoadedAssemblies != null && LazyLoadedAssemblies.Any(a => Path.GetFileNameWithoutExtension(a.ItemSpec) == fileName);
```
Comment (javiercn) [inline src/Components/WebAssembly/Sdk/src/GenerateBlazorWebAssemblyBootJson.cs:152]: I think we want to either require the extension to be specified or not to be specified at all, I'm not sure if we want to allow both options (with or without the file extension)

**Comment by adrianwright109** [comment]
I'm not 100% sure this is the correct fix.

I had trouble debugging the MSBuild task `GenerateBlazorWebAssemblyBootJson.cs`.  In the end I found a Stack overflow answer which showed I could add some debug code which would allow me time to attach a VS debugger to the `dotnet.exe` process and then step through.

```
#if DEBUG
            try
            {
                Console.WriteLine("Waiting for debugger to attach.");
                Console.WriteLine($"Process ID: {System.Diagnostics.Process.GetCurrentProcess().Id}");

                while (!System.Diagnostics.Debugger.IsAttached)
                {
                    System.Threading.Thread.Sleep(100);
                }

                System.Diagnostics.Debugger.Break();
            }
            catch (Exception ex)
            {
                Log.LogErrorFromException(ex);
                return false;
            }
#endif
```

With this fix I think that below formats below will work:
`<BlazorWebAssemblyLazyLoad Include="RazorClassLibrary" />`
and
`<BlazorWebAssemblyLazyLoad Include="RazorClassLibrary.dll" />`

It may not be desirable to allow both ?

I'm not sure how the value in the `Include` attribute makes its way through to the MSBuild task and into the LazyLoadedAssemblies `.ItemSpec` property.

**Comment by javiercn** [comment]
> I'm not sure how the value in the `Include` attribute makes its way through to the MSBuild task and into the LazyLoadedAssemblies `.ItemSpec` property.

It's an element in an ItemGroup (kind-of like an array in MSBuild). The ItemSpec represents the value on the "Include" and then it can have additional metadata associated with it.

> It may not be desirable to allow both ?

It's a matter of how you explain people what to put in there. If you force the extension to be present, it is clear that you need to refer to the file name. If you don't have an extension at all, then it's not clear whether you are referring to the file name (and assuming the extension) or if you are referring to, for example, the assembly name, package name or something else (which might be different from the file name).

Requiring the extension all the time makes the answer clear and straightforward and the errors easy to diagnose, since there is a single valid value you can use per assembly.

As an aside question for @captainsafia do we produce an error on the build if we don't match a lazy loaded assembly definition with a file on the output? We should produce an error in this case, as something you've explicitly declared is not doing what you expect.

File: src/Components/WebAssembly/Sdk/src/GenerateBlazorWebAssemblyBootJson.cs (hunk 1 above)
Comment (campersau) [reply src/Components/WebAssembly/Sdk/src/GenerateBlazorWebAssemblyBootJson.cs:152]: I guess you could change line 91 to this.

```diff
- if (IsLazyLoadedAssembly(fileName))
+ if (IsLazyLoadedAssembly(resourceName))
```

**Comment by captainsafia** [comment]
> It may not be desirable to allow both ?

Yes. We'd only like to support `.dll` filenames so that it is consistent with what our `LazyAssemblyLoader` API takes.

> As an aside question for @captainsafia do we produce an error on the build if we don't match a lazy loaded assembly definition with a file on the output?

We don't produce a build error but we do produce a runtime error.

**Comment by javiercn** [comment]
> We don't produce a build error but we do produce a runtime error.

@captainsafia Is there a reason why we don't produce a build error? At the time we are building blazor.boot.json we already know the shape of the final app (at least the dlls and their contents), don't we? So we should produce an error at that point if we detect a LazyLoadAssembly entry that doesn't match any dll instead of waiting for the app to run.

I'm by no means asking this PR addresses that, but we should fix this before we RTM, so consider filing a separate issue for it?

**Review by captainsafia — APPROVED** [review]
Thanks for submitting this fix, @adrianwright109!

@guardrex We'll need to update the lazy-loading docs for RC1 to include extension names in the `Include` for `BlazorWebAsssemblyLazyLoad`.

**Pull request merged**